### PR TITLE
Automagic get for open and h5py.File 'r' operations

### LINKED ISSUE #155: "fuse" like operation when used a Python module
We could provide a basic overload of "open" (and may be some of os.path) to automagically fetch requested files (e.g. upon `import datalad.magicio` or smth like that).  Use-case -- when running some tutorial, which opens (nibabel) files, we could check if file_has_content and if not -- fetch it only then passing call to stock 'open'.  Even though it might be fragile etc it might be a silver bullet working in 9x% of cases. TODO: verify operation in conjunction with nipype ;)  In comparison to proper fuse overlay, this one would be OS-agnostic.

### PR DESCRIPTION
**Pull request by yarikoptic**
I guess naming could be better --suggestions?

I also thought to implement mode of operation as "python -m datalad" which would enable AutomagicIO but its range of use-cases is probably minimal so "not sure".  We better allow for something like `datalad.enable_autoget()`?

I haven't yet tried on some real example (e.g. pymvpa tutorial) but it should work (tests work! ;) )
Close #155

### CODE REVIEW

**Comment by yarikoptic** [comment]
FYI buildbots are failing due to problems with datalad.org DNS. IIRC that DNS query was refactored away anyways in #203 , so look into that one first

File: datalad/auto.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+from .support.annexrepo import AnnexRepo
+from .support.gitrepo import GitRepo
+from .support.exceptions import CommandError
+from .cmd import Runner
+
+from .utils import swallow_outputs
+lgr = logging.getLogger("datalad.auto")
+
+
+# TODO: shouldn't be some classmethod of GitRepo?
```
Comment (bpoldrack) [inline datalad/auto.py]: What about using GitRepo.get_indexed_files()?

**Comment by bpoldrack** [comment]
To be honest: While this is theoretically nice to have, I have no idea when to use this functionality. For now there's no command that would need to read an annexed file, as far as I'm aware. It only matters for tests, that check for correct content, right?

File: datalad/auto.py (hunk 1 above)
Comment (yarikoptic) [inline datalad/auto.py]: but I don't want all indexed files (of which there could be thousands) -- I just need a quick check if a given file is under git.  But I see now where this wouldn't work -- direct mode (will fix for that ) :-/  need to go through Annex straight then **But** then #204 is in the way -- there really should be a "read-only" mode of operation on annex, where at least us directly not doing anything directly (like init) -- most of the  annex commands could indeed do their 'init' stage if needed though

File: datalad/support/gitrepo.py [hunk 2]
```diff
@@ -69,7 +72,10 @@ def _normalize_path(base_dir, path):
     # But we don't want to realpath relative paths, in case cwd isn't the correct base.
 
     if isabs(path):
-        path = realpath(path)
+        # path might already be a symlink pointing to annex etc,
+        # so realpath only its directory, to get "inline" with realpath(base_dir)
+        # above
+        path = opj(realpath(dirname(path)), basename(path))
```
Comment (yarikoptic) [inline datalad/support/gitrepo.py]: @bpoldrack this was a generic fix btw -- just didn't bother for a separate PR, if this one drags too long, we might want to pick it up separately

**Comment by yarikoptic** [comment]
Indeed, I better implement `__main__` -- pushed now.  And indeed demo is better than a thousand of words:
```
$> git clone http://data.pymvpa.org/datasets/tutorial_data/.git
Cloning into 'tutorial_data'...
Checking connectivity... done.

$> wget https://raw.githubusercontent.com/PyMVPA/PyMVPA/master/doc/examples/hyperalignment.py
...
2015-09-15 08:35:42 (267 MB/s) - ‘hyperalignment.py’ saved [17408/17408]

$> MVPA_LOCATION_TUTORIAL_DATA=/tmp/tutorial_data python hyperalignment.py 
 Loading data...
Traceback (most recent call last):
  File "hyperalignment.py", line 51, in <module>
    ds_all = h5load(filepath)
  File "/usr/lib/python2.7/dist-packages/mvpa2/base/hdf5.py", line 850, in h5load
    hdf = h5py.File(filename, 'r')
  File "/usr/lib/python2.7/dist-packages/h5py/_hl/files.py", line 260, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)
  File "/usr/lib/python2.7/dist-packages/h5py/_hl/files.py", line 89, in make_fid
    fid = h5f.open(name, flags, fapl=fapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/build/h5py-TT07Di/h5py-2.5.0/h5py/_objects.c:2467)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/build/h5py-TT07Di/h5py-2.5.0/h5py/_objects.c:2424)
  File "h5py/h5f.pyx", line 76, in h5py.h5f.open (/build/h5py-TT07Di/h5py-2.5.0/h5py/h5f.c:1801)
IOError: Unable to open file (Unable to open file: name = '/tmp/tutorial_data/hyperalignment_tutorial_data.hdf5.gz', errno = 2, error message = 'no such file or directory', flags = 0, o_flags = 0)

$> MVPA_LOCATION_TUTORIAL_DATA=/tmp/tutorial_data python -m datalad hyperalignment.py 
 Loading data...
/tmp/tutorial_data/.git/annex/tmp/SHA256E-s1576 100%[=========================================================================================================>]  15.04M  11.2MB/s   in 1.3s   
  10 subjects
  Per-subject dataset: 56 samples with 3509 features
  Stimulus categories: Chair, DogFace, FemaleFace, House, MaleFace, MonkeyFace, Shoe
 Performing classification analyses...
  within-subject.../usr/lib/python2.7/dist-packages/numpy/lib/utils.py:134: DeprecationWarning: `fprob` is deprecated!
fprob is deprecated in scipy 0.14, use stats.f.sf or special.fdtrc instead

  warnings.warn(depdoc, DeprecationWarning)
 done in 1.4 seconds
  between-subject (anatomically aligned)...done in 0.9 seconds
  between-subject (hyperaligned)...done in 4.0 seconds
 Average classification accuracies:
...
```
Now scale this to carrying out some analysis over a sweep of annex'es without knowing what data files you really need to have upfront (or just being time consuming to select them to download)

**Comment by yarikoptic** [comment]
To decrease overhead this feature probably needs quite a bit of tuning and may be allow to specify a selection of paths to monitor, but that is later I guess...  as for direct mode -- I indeed would need just to wait until we decide on #204, strip out check for being under Git, we would need AnnexRepo smarter to be able to initialize/discover the root directory for the annex, and then it should work in direct mode as well.  Meanwhile I will leave it in this state for digestion/feedback/acceptance ;)

File: datalad/auto.py (hunk 1 above)
Comment (bpoldrack) [inline datalad/auto.py]: Hm. May be worth measuring whether a ls-files call is actually faster than checking indexed_files.

Regarding the "read-only" mode: Any git-annex call will execute an init if it wasn't done before, as far as I'm aware. So, what exactly do you expect to be different in that mode?

Comment (yarikoptic) [inline datalad/auto.py]: my timing on smaug in a repo with >10,000 files:
```shell
smaug:/mnt/scrap/datalad/mimic_repos/hcp-ls.20141020-500subjects/144226/hcp-openaccess/500subject/144226/MNINonLinear
$> time git ls-files wmparc.nii.gz --error-unmatch
wmparc.nii.gz
git ls-files wmparc.nii.gz --error-unmatch  0.01s user 0.00s system 95% cpu 0.013 total
```
try to beat that one? ;) ok:
```
smaug:/mnt/scrap/datalad/mimic_repos/hcp-ls.20141020-500subjects/144226
$> python -c "import time; from datalad.support.annexrepo import GitRepo; annex=GitRepo('$PWD'); t1=time.time(); annex.get_indexed_files(); print(time.time() - t1)"
0.0757648944855
```
BUT you are correct that there might be (more) appropriate function within GitRepo. The problem though is that for that one you need to first figure out top directory which has .git, whenever ls-files  can work from wherever

re read-only -- lets' continue in #204?

Comment (bpoldrack) [inline datalad/auto.py]: Well, ls-files has to somehow find git's index too, hasn't it? So, finding the top is a task, that ls-files has to accomplish as well, I guess. Anyway, as long as we don't find a faster search on the index of a GitPython.Repo, let's stay with ls-files.

**Comment by yarikoptic** [comment]
ok @bpoldrack -- addressed my concern in 57de57bbf71275bd06eaf2080a100264a9ef05ff which is now should be part of this PR... or may be it is time to extract those fixups in to another branch unless this one gets merged ;)

**Comment by yarikoptic** [comment]
ok -- extracted relevant commits to #218 with some whining/discussion attached ;-)

**Comment by hanke** [comment]
Yarik promised that #204 is no issue in this context anymore. Merging.

**Pull request merged**